### ISSUE TITLE
Consume before publish prevents consuming any messages?

### ISSUE THREAD

**Issue by jeffparsons** [sozu-proxy/lapin#65]
I've run into some confusion around a consumer intermittently not consuming any messages. Thankfully I was able to reproduce it in a really simple test case (so I can leave the rest of my program's code behind), by turning https://github.com/sozu-proxy/lapin/blob/master/futures/examples/client.rs "upside-down" such that it starts trying to consume before publishing any message.

The end result looks pretty simple, so I'm guessing this is me misunderstanding something basic about how to use the library. But if that's the case, hopefully there's some way to detect whatever silly thing I'm doing, or some way to guide other travellers away from my mistake.

Here's the code:

```rust
extern crate lapin_futures as lapin;
extern crate futures;
extern crate tokio_core;

use futures::future::Future;
use futures::Stream;
use tokio_core::reactor::Core;
use tokio_core::net::TcpStream;
use lapin::types::FieldTable;
use lapin::client::ConnectionOptions;
use lapin::channel::{BasicConsumeOptions,BasicPublishOptions,BasicProperties,ExchangeBindOptions,ExchangeUnbindOptions,ExchangeDeclareOptions,ExchangeDeleteOptions,QueueBindOptions,QueueDeclareOptions};
use std::thread;

fn main() {
    let mut core = Core::new().unwrap();

    let handle = core.handle();
    let addr = "192.168.50.100:5672".parse().unwrap();

    core.run(
        TcpStream::connect(&addr, &handle).and_then(|stream| {
            lapin::client::Client::connect(stream, &ConnectionOptions {
                username: "guest".to_string(),
                password: "guest".to_string(),
                frame_max: 65535,
                ..Default::default()
            })
        }).and_then(|(client, heartbeat_future_fn)| {
            let heartbeat_client = client.clone();
            thread::Builder::new().name("heartbeat thread".to_string()).spawn(move || {
                Core::new().unwrap().run(heartbeat_future_fn(&heartbeat_client)).unwrap();
            }).unwrap();

            client.create_channel().and_then(|channel| {
                let id = channel.id;
                println!("created channel with id: {}", id);

                let ack_channel = channel.clone();
                let publish_channel = channel.clone();

                channel.queue_declare("hello", &QueueDeclareOptions::default(), &FieldTable::new()).and_then(move |_| {
                    println!("channel {} declared queue {}", id, "hello");

                    handle.spawn(
                        channel.basic_consume("hello", "my_consumer", &BasicConsumeOptions::default(), &FieldTable::new()).and_then(|stream| {
                            println!("got consumer stream");

                            stream.for_each(move |message| {
                                println!("got message: {:?}", message);
                                println!("decoded message: {:?}", std::str::from_utf8(&message.data).unwrap());
                                ack_channel.basic_ack(message.delivery_tag)
                            })
                        }).map_err(|err| {
                            println!("uh oh: {:?}", err);
                            ()
                        })
                    );
                    futures::future::ok::<(), std::io::Error>(())
                }).and_then(move |_| {
                    publish_channel.exchange_declare("hello_exchange", "direct", &ExchangeDeclareOptions::default(), &FieldTable::new()).and_then(move |_| {
                        publish_channel.queue_bind("hello", "hello_exchange", "hello_2", &QueueBindOptions::default(), &FieldTable::new()).and_then(move |_| {
                            publish_channel.basic_publish(
                                "hello_exchange",
                                "hello_2",
                                b"hello from tokio",
                                &BasicPublishOptions::default(),
                                BasicProperties::default().with_user_id("guest".to_string()).with_reply_to("foobar".to_string())
                            ).map(|confirmation| {
                                println!("publish got confirmation: {:?}", confirmation)
                            }).and_then(move |_| {
                                publish_channel.exchange_bind("hello_exchange", "amq.direct", "test_bind", &ExchangeBindOptions::default(), &FieldTable::new()).and_then(move |_| {
                                        publish_channel.exchange_unbind("hello_exchange", "amq.direct", "test_bind", &ExchangeUnbindOptions::default(), &FieldTable::new()).and_then(move |_| {
                                                publish_channel.exchange_delete("hello_exchange", &ExchangeDeleteOptions::default()).and_then(move |_| {
                                                        publish_channel.close(200, "Bye")
                                                })
                                        })
                                })
                            })
                        })
                    })
                })
            })
        })
    ).unwrap();

    core.run(futures::future::empty::<(), ()>()).unwrap();
}
```

If I'm lucky, then it'll print this:

```
created channel with id: 1
channel 1 declared queue hello
got consumer stream
got message: Message { delivery_tag: 1, exchange: "hello_exchange", routing_key: "hello_2", redelivered: true, properties: Properties { content_type: None, content_encoding: None, headers: None, delivery_mode: None, priority: None, correlation_id: None, reply_to: Some("foobar"), expiration: None, message_id: None, timestamp: None, type_: None, user_id: Some("guest"), app_id: None, cluster_id: None }, data: [104, 101, 108, 108, 111, 32, 102, 114, 111, 109, 32, 116, 111, 107, 105, 111] }
decoded message: "hello from tokio"
```

But lots of the time it'll print this...

```
created channel with id: 1
channel 1 declared queue hello
got consumer stream
publish got confirmation: None
```

...and then make no further progress, with the message sitting in the queue until I kill my program and it gets auto-deleted.

Is it obvious from looking at this that I'm doing something wrong? Perhaps something to do with `spawn`ing my consumer?

**Comment by Keruspe** [comment]
Hi,

You shouldn't use the "same" channel in multiple threads. RabbitMQ really discourages that.
Instead of using one of the clones of the channel you created in your spawn, could you try creating another channel for that?

**Comment by Keruspe** [comment]
Also, I'm not familiar with handle.spawn but I guess what it says is "spawn this future when you can, then resume what you were doing", so it just depends on whether tokio spawns it before or after the publish part. As it never ends, in the former case the publish is never done.
Care to try within a new thread instead? (like for the heartbeat stuff at the beginning)

**Comment by michaelklishin** [comment]
Also see server logs for any possible channel or connection-level exceptions and consider [taking a traffic capture](https://www.rabbitmq.com/amqp-wireshark.html) to see what is really going on.

Consuming and publishing on a channel is perfectly fine as far as the protocol goes but channel sharing across threads (or similar) is a no-no for publishing and should be considered with great care in general. Of course, what exactly is safe to do varies from client to client.

**Comment by jeffparsons** [comment]
Thanks @Keruspe and @michaelklishin for your replies! 😃 

> You shouldn't use the "same" channel in multiple threads. RabbitMQ really discourages that.Instead of using one of the clones of the channel you created in your spawn, could you try creating another channel for that?

I've changed my code to be more like the example, by using one "fresh" channel for the publisher side of things, and another for the consumer side. (Only cloning the consumer channel for ack, as the example does.)

It now looks like this:

```rust
extern crate lapin_futures as lapin;
extern crate futures;
extern crate tokio_core;

use futures::future::Future;
use futures::Stream;
use tokio_core::reactor::Core;
use tokio_core::net::TcpStream;
use lapin::types::FieldTable;
use lapin::client::ConnectionOptions;
use lapin::channel::{BasicConsumeOptions,BasicPublishOptions,BasicProperties,ExchangeBindOptions,ExchangeUnbindOptions,ExchangeDeclareOptions,ExchangeDeleteOptions,QueueBindOptions,QueueDeclareOptions};
use std::thread;

fn main() {
    let mut core = Core::new().unwrap();

    let handle = core.handle();
    let addr = "192.168.50.100:5672".parse().unwrap();

    core.run(
        TcpStream::connect(&addr, &handle).and_then(|stream| {
            lapin::client::Client::connect(stream, &ConnectionOptions {
                username: "guest".to_string(),
                password: "guest".to_string(),
                frame_max: 65535,
                ..Default::default()
            })
        }).and_then(|(client, heartbeat_future_fn)| {
            let heartbeat_client = client.clone();
            thread::Builder::new().name("heartbeat thread".to_string()).spawn(move || {
                Core::new().unwrap().run(heartbeat_future_fn(&heartbeat_client)).unwrap();
            }).unwrap();

            client.create_channel().and_then(|channel| {
                let id = channel.id;
                println!("created channel with id: {}", id);

                let ack_channel = channel.clone();

                channel.queue_declare("hello", &QueueDeclareOptions::default(), &FieldTable::new()).and_then(move |_| {
                    println!("channel {} declared queue {}", id, "hello");

                    handle.spawn(
                        channel.basic_consume("hello", "my_consumer", &BasicConsumeOptions::default(), &FieldTable::new()).and_then(|stream| {
                            println!("got consumer stream");

                            stream.for_each(move |message| {
                                println!("got message: {:?}", message);
                                println!("decoded message: {:?}", std::str::from_utf8(&message.data).unwrap());
                                ack_channel.basic_ack(message.delivery_tag)
                            })
                        }).map_err(|err| {
                            println!("uh oh: {:?}", err);
                            ()
                        })
                    );
                    futures::future::ok::<(), std::io::Error>(())
                }).and_then(move |_| {
                    client.create_channel().and_then(|publish_channel| {
                        let id = publish_channel.id;
                        println!("created channel with id: {}", id);

                        publish_channel.exchange_declare("hello_exchange", "direct", &ExchangeDeclareOptions::default(), &FieldTable::new()).and_then(move |_| {
                            publish_channel.queue_bind("hello", "hello_exchange", "hello_2", &QueueBindOptions::default(), &FieldTable::new()).and_then(move |_| {
                                publish_channel.basic_publish(
                                    "hello_exchange",
                                    "hello_2",
                                    b"hello from tokio",
                                    &BasicPublishOptions::default(),
                                    BasicProperties::default().with_user_id("guest".to_string()).with_reply_to("foobar".to_string())
                                ).map(|confirmation| {
                                    println!("publish got confirmation: {:?}", confirmation)
                                }).and_then(move |_| {
                                    publish_channel.exchange_bind("hello_exchange", "amq.direct", "test_bind", &ExchangeBindOptions::default(), &FieldTable::new()).and_then(move |_| {
                                            publish_channel.exchange_unbind("hello_exchange", "amq.direct", "test_bind", &ExchangeUnbindOptions::default(), &FieldTable::new()).and_then(move |_| {
                                                    publish_channel.exchange_delete("hello_exchange", &ExchangeDeleteOptions::default()).and_then(move |_| {
                                                            publish_channel.close(200, "Bye")
                                                    })
                                            })
                                    })
                                })
                            })
                        })
                    })
                })
            })
        })
    ).unwrap();

    core.run(futures::future::empty::<(), ()>()).unwrap();
}
```

Unfortunately the intermittent hanging is the same as before.

> Also, I'm not familiar with handle.spawn but I guess what it says is "spawn this future when you can, then resume what you were doing", so it just depends on whether tokio spawns it before or after the publish part. As it never ends, in the former case the publish is never done.

My understanding is that `Handle::spawn` just enqueues a task, along with any others that might be current, and once all other tasks in line have yielded (said to the reactor that they can't currently make progress) then it would get a turn, and so on. So no task should block any other task unless, e.g., one of them has taken out a lock and not yielded the lock when it realised it couldn't make any progress, or something like that. (I.e. anything that would cause a deadlock in any other kind of concurrent programming.)

> Care to try within a new thread instead? (like for the heartbeat stuff at the beginning)

As in, running a completely separate reactor (`Core` in Tokio parlance) for it? I don't doubt that this would work (I will try, to verify this!), but my understanding is that this shouldn't be necessary for any futures-based API; if a given [task](https://docs.rs/futures/0.1.14/futures/task/struct.Task.html) isn't able to make progress, then I thought it should avoid holding any locks / etc. that another task on the same reactor might need to make progress — i.e. they shouldn't block or otherwise interfere with each other. Indeed I thought one of the core principles of futures-rs/Tokio is that you should be able to share one reactor for pretty much everything. Maybe I'm mistaken...

As for the heartbeat stuff, my understanding was that it's only done on a separate thread in case you accidentally block the reactor thread for a long time by (inappropriately!) doing a bunch of CPU-bound work on it, so the heartbeat timer misses a beat. That is to say, I didn't think it was _necessary_, but only best practice to avoid accidents.

I'll try splitting the consumer off to a separate reactor and report back.

Thanks again! 😁

**Comment by jeffparsons** [comment]
@michaelklishin Thanks for the tip re Wireshark's AMQP support; I'll look into it!

**Comment by michaelklishin** [comment]
Are there any channel-level or connection-level exceptions in the server log? The former close the channel (so there can be no more activity on it) and the latter close the connection.

**Comment by Keruspe** [comment]
The consumer is a never-ending stream

**Comment by jeffparsons** [comment]
@michaelklishin I will get around to taking a look at Rabbit logs, and hooking up Wireshark. I just want to dump some stuff here first, if that's ok, to get my head around what behaviour I should be expecting in an ideal world so I know what I'm looking for, and to explain where I'm coming from.

> The consumer is a never-ending stream

@Keruspe I think we're not quite on the same page, because I'm not sure what is the consequence of it being a never-ending stream. My understanding is that even though it's an infinite stream, it shouldn't block the reactor from continuing with other work _if it can't make any progress_ when it is polled.

To clarify what I mean, the following example spawns two infinite streams, each of which can _always_ make progress, because they don't do any IO.

```rust
extern crate futures;
extern crate tokio_core;

use tokio_core::reactor::Core;

fn main() {
    use futures::Stream;

    let mut core = Core::new().unwrap();
    let handle = core.handle();

    println!("Spawning infinite stream A...");
    let iter_a = (0..std::usize::MAX).map(|i| Ok(i));
    handle.spawn(futures::stream::iter(iter_a).for_each(|i| {
        println!("A: {}", i);
        futures::future::ok::<(), ()>(())
    }));

    println!("Spawning infinite stream B...");
    let iter_b = (0..std::usize::MAX).map(|i| Ok(i));
    handle.spawn(futures::stream::iter(iter_b).for_each(|i| {
        println!("B: {}", i);
        futures::future::ok::<(), ()>(())
    }));

    core.run(futures::future::empty::<(), ()>()).unwrap();
}
```

This outputs:

```
Spawning infinite stream A...
Spawning infinite stream B...
A: 0
A: 1
A: 2
A: 3
A: 4
A: 5
A: 6
A: 7
A: 8
A: 9
A: 10
A: 11
A: 12
```

(Etc.)

This makes sense to me. Because we spawn A first, and it is _always ready_ (i.e. `poll()` will never return `Async::NotReady`), the reactor will never park it and move on to other tasks, and B will never get a turn.

On the other hand, here's an example where both streams will sometimes not be able to make progress, because they are waiting on IO events. My intent in this example is to model something closer to what's going on under the hood in talking AMQP; we're dealing with infinite streams that have to wait for IO.

```rust
extern crate futures;
extern crate tokio_core;
extern crate tokio_io;

use tokio_core::reactor::{ Core, Handle };

use futures::{Future, Stream};
use tokio_io::{io, AsyncRead};
use tokio_core::net::TcpListener;

fn main() {
    let mut core = Core::new().unwrap();

    let server1 = mk_server(core.handle(), "Server 1", "127.0.0.1:12345");
    let server2 = mk_server(core.handle(), "Server 2", "127.0.0.1:12346");

    // Spawn both servers. Note that both resolve to an endless stream
    // of connections, so if `for_each` on a stream inherently blocks
    // a reactor, then only one of these servers should be able to accept connections.
    println!("Spawning Server 1...");
    core.handle().spawn(server1);
    println!("Spawning Server 2...");
    core.handle().spawn(server2);

    core.run(futures::future::empty::<(), ()>()).unwrap();
}

fn mk_server(handle: Handle, name: &'static str, addr: &'static str) -> Box<Future<Item=(), Error=()>> {
    use futures::IntoFuture;

    let addr = addr.parse().unwrap();
    let tcp = TcpListener::bind(&addr, &handle).unwrap();

    let f = tcp.incoming().for_each(move |(tcp, _)| {
        println!("{} got a connection!", name);

        let (reader, writer) = tcp.split();

        // Future of the copy
        let bytes_copied = io::copy(reader, writer);

        // ... after which we'll print what happened
        let handle_conn = bytes_copied.map(move |(n, _, _)| {
            println!("{} wrote {} bytes", name, n)
        }).map_err(move |err| {
            println!("{} IO error {:?}", name, err)
        });

        // Spawn the future as a concurrent task
        handle.spawn(handle_conn);

        Ok(())
    }).into_future().map_err(|_| ());
    Box::new(f)
}
```

If I open connections (`telnet localhost 12345`, `telnet localhost 12346`, repeat in random order) this outputs:

```
Spawning Server 1...
Spawning Server 2...
Server 1 got a connection!
Server 1 wrote 17 bytes
Server 1 got a connection!
Server 1 wrote 0 bytes
Server 2 got a connection!
Server 2 wrote 9 bytes
Server 1 got a connection!
Server 1 wrote 9 bytes
Server 2 got a connection!
Server 2 wrote 17 bytes
```

So even though both server futures are driving infinite streams, and are both spawned on the same reactor, both are able to make progress. This is closer to the behaviour I would've expected by default in Lapin when consuming messages.

Now, all that said, I'm still _extremely new_ to this Tokio stuff, so I'm still assuming the most likely explanation is that I'm confused about something. But I figured it'd be worth dropping these examples here to clarify exactly what it is that I'm confused about!

Does this make any sense now? 😄 

Thanks for your patience!

**Comment by Keruspe** [comment]
Can you run your lapin example with RUST_LOG=trace ?

**Comment by jeffparsons** [comment]
```
[jeff@jrti:~/Desktop/lapin-confusion]$ RUST_LOG=trace target/debug/lapin-confusion > unlucky-stdout.txt 2> unlucky-stderr.txt
^C
[jeff@jrti:~/Desktop/lapin-confusion]$ RUST_LOG=trace target/debug/lapin-confusion > lucky-stdout.txt 2> lucky-stderr.txt
^C
```
[lucky-stdout.txt](https://github.com/sozu-proxy/lapin/files/1161700/lucky-stdout.txt)
[lucky-stderr.txt](https://github.com/sozu-proxy/lapin/files/1161701/lucky-stderr.txt)
[unlucky-stdout.txt](https://github.com/sozu-proxy/lapin/files/1161702/unlucky-stdout.txt)
[unlucky-stderr.txt](https://github.com/sozu-proxy/lapin/files/1161703/unlucky-stderr.txt)
[main.rs.txt](https://github.com/sozu-proxy/lapin/files/1161705/main.rs.txt)

**Comment by Keruspe** [comment]
Hmm, I did some testing locally, it seems that the message is correctly published, is correctly handled under the hood, but the consumer is never polled.
I see in the logs the the consumer gets polled a first time, returning NotReady, then the publishing is done, the messages is received by lapin-async, but the consumer never gets polled again.
You can track this by looking for `TRACE:lapin_futures::consumer: consumer[my_consumer] poll` in the logs

**Comment by jeffparsons** [comment]
@Keruspe Looking through the Lapin source, I can't figure out what would actually trigger readiness if the future was ever `Async::NotReady`.

Compare with the `TcpListener` implementation in Tokio, which [uses `PollEvented`](https://tokio-rs.github.io/tokio-core/src/tokio_core/net/tcp.rs.html#144) to make sure that the task will be polled again eventually if it returns `NotReady`. If I understand correctly, then `PollEvented` registers interest in events from the underlying IO object, which will cause the reactor to poll the `TcpListener` whenever the underlying IO object becomes readable again (even if it doesn't necessarily allow the current task to make progress; it can't know that).

I can't see where the equivalent mechanism is implemented in `lapin_futures`. Am I just missing something here, or is that behaviour simply missing?

**Comment by jeffparsons** [comment]
In case you're interested, I've tidied up my test case a bit, made it deterministic (well, unless starting the consumer stream takes over a second, that is! 😁) and pushed a branch with it here: https://github.com/jeffparsons/lapin/commit/dc776b63edad79ee17f5d7ec915014ed5f8f0582

**Comment by Keruspe** [comment]
I'm going to give it a look

**Comment by jeffparsons** [comment]
Did you manage to glean any further insight into what's going on here? If you've got other priorities at the moment, then I'd be happy to take a swing.

**Comment by Keruspe** [comment]
Hi,

I have some WIP but couldn't get to an acceptable solution yet.

You can see my WIP here: https://github.com/Keruspe/lapin/commit/f19720c4b27fcb8285757fdabbf66bad57f72560

**Comment by Keruspe** [comment]
You can try to get it working of course, that would be greatly appreciated!

**Comment by jeffparsons** [comment]
I had a go at figuring this out the other day, and again today, and I'm still bashing my head against it without really understanding what's going on. The behaviour I'm seeing on your branch is this:

- The future returned from `exchange_declare` gets polled.
- Upstream poll returns `NotReady`.
- Consumer gets polled.
- Among other things, upstream poll finds `Exchange(DeclareOk(DeclareOk)))`, and inserts the response into `Connection::finished_reqs`, ready for next time the exchange declare future is polled.
- But then the exchange declare future is never polled again.

This _looks_ as though the task that `exchange_declare` was called on is _never_ being un-parked after the initial poll, even though the attempt to read from the underlying `AsyncRead` _should_ have scheduled it to be unparked when it becomes readable again. That seems like a _really unlikely_ explanation (or surely more people would have run a similar issue), so the next thing I'm going to try to look at is exactly when the relevant tasks are being parked and un-parked. I might have to make a branch of tokio to expose task IDs to help with this. Maybe understanding that will point me in a more useful direction.

This is turning out to be harder to understand than I had expected!

**Comment by jeffparsons** [comment]
For the record, forcing the task to be notified with a timeout allows everything to continue happily. So the problem definitely involves the task not being notified of IO readiness, but the question is _why_.

```rust
let the_rest = client.create_channel().and_then(|publish_channel| {
    // ...
});

// Notify the task after a short while.
{
    use futures::task;
    use std::time::Duration;
    use tokio_core::reactor::Timeout;
    let current_task = task::current();
    let f = Timeout::new(Duration::new(1, 0), &handle2).unwrap()
        .map_err(|err| {
            panic!("timeout failed: {:?}", err);
        })
        .map(move |timeout| {
            current_task.notify();
        });
    handle2.spawn(f);
}

the_rest
```

**Comment by jeffparsons** [comment]
I haven't looked at this in a long time and (to be realistic) probably won't be getting back into the guts of it any time soon. But I did have a thought a while back that might be worth considering:

What about ripping out all the current glue between the futures layer and the underlying stuff going on in `lapin_async`, and instead use [one-shot channels](http://alexcrichton.com/futures-rs/futures/sync/oneshot/index.html) wherever possible (storing the sender in a dictionary like the one you have for pending results at the moment — and similar for streams)? That way you could avoid carrying your own complexity of trying to make sure things are polled at the right time, and be confident that if something makes it to the `lapin_async` layer, it's probably going to get communicated up to the relevant future/stream interface properly.

I assume there'd be some kind of performance hit, but it's likely to be dwarfed by the IO involved, and I'm generally pretty keen on "make it work, then make it fast" anyway. 😁 

Again, I haven't looked at this in a while so I can't remember how neatly it would fit in to your implementation, but my gut says it'd be worth exploring.

**Comment by pimeys** [comment]
Was there any resolution for this? Having an async rabbitmq client would be enormous gain for us, but I'm having the exact same problem as described in this issue.

I need to consume messages and then send a response to another exchange when the work is done. I could run another client for publishing in another thread and use channels for communication between the consumer and producer, but that would make the codebase overly complex and I should be able to produce from the same client I'm consuming from.

What I can do is I can send 100 test messages to the queue, then start my consumer. The messages are consumed and sent to third party, after which they don't consume any new messages and I get `NotReady` to the logs.

**Comment by Keruspe** [comment]
Is this bug still relevant with current master?

**Comment by jeffparsons** [comment]
I haven't tried recently, and would need to install RabbitMQ do so. Do you believe there are fixes on master that _should_ fix this class of problem? If so, I'd be happy to try it out again and create a PR that tests the behaviour.

Otherwise, I'll probably wait for the "futures 0.3" work to land upstream to have another play with this; I'm not really interested in checking it out on the off chance that it might have _incidentally_ been fixed by some other work, and if there is still work to be done to make this behave correctly then maybe there'll be some changes in "future 0.3" that make it easier to get this stuff right. (I don't actually know whether that will be the case — just wanting to avoid expending effort now if the whole futures ecosystem might change drastically in the near future.)

**Comment by Keruspe** [comment]
There has been some recent fix to force polling some futures which could fix this issue yes (not yet in a release)

**Comment by jeffparsons** [comment]
I merged the latest master changes into [my branch with the test](https://github.com/jeffparsons/lapin/tree/test_consumer_not_ready_polls_again). It hangs forever.

I tried moving the timeout to be before consuming instead of before publishing... and now the message is received successfully.

So I believe the bug still exists.

**Comment by jeffparsons** [comment]
@Keruspe BTW, is there anything holding up https://github.com/sozu-proxy/lapin/pull/93? I think after that lands I'll rewrite my test to be mergeable (i.e. not just hang forever if it doesn't work 😅) and be really easy to follow with descriptive panic messages that explain exactly what went wrong, etc.

**Comment by Keruspe** [comment]
#93 is now merged

**Comment by jgallagher** [comment]
I believe I bumped into this same issue today and think I understand what's going. It looks like the root cause of this is sharing the same underlying `TcpStream` across multiple tasks (which isn't normally possible, but `Client` wraps it in an `Arc<Mutex<...>>`). I think the order of events is roughly:

* Start running the "main" future (`core.run(...)`). Call this Task A.
* Spawn a new task to handle `basic_consume` - call this Task B.
* Task B tries to read on the TCPStream - there's nothing there, so it parks and should be woken once there is data available.
* Task A starts running again and (eventually) gets blocked on the TCPStream for some reason. Now it parks and wants to be woken once the TCPStream is ready.

At this point you conceptually have two tasks parked on the same TCPStream, which AFAICT isn't supported by tokio - the later one overwrites the earlier one. Moving forward:

* Data arrives on the TCPStream that's intended for Task B.
* Task _A_ gets woken up, because it was the last one to park on the stream.
* Task A can't make any progress with the data that showed up, so it just parks again.
* Task B is never woken.

I think this means that the current implementation of channels (and maybe clients?) can't really be used with multiple tasks on the same event loop. I _think_ it might still be safe to use them across multiple threads, as long as each thread only has one task using it on its event loop, but haven't tried that to confirm.

**Comment by jeffparsons** [comment]
I have a test for this. Travelling right now but should be home in a few
days; will PR when back.

**Comment by jeffparsons** [comment]
Test for this is here: https://github.com/sozu-proxy/lapin/pull/110 — hopefully easy to switch around to try different combinations of things.

**Comment by Keruspe** [comment]
I'm pretty sure #116 will fix this, it'll get merged soon, can you retry once it's done?

**Comment by jeffparsons** [comment]
This appears to be fixed by recent changes. I had some tests that showed this working, but don't have a "real" project anymore to test this with.
